python/feathub/processors/local/sliding_window_utils.py

```diff
@@ -0,0 +1,353 @@
+        df_copy[EVENT_TIME_ATTRIBUTE_NAME] * 1000
+    ).astype(np.int)
+
+    if len(window_descriptor.group_by_keys) > 0:
+        agg_df = df_copy.groupby(by=window_descriptor.group_by_keys).apply(
+            _get_sliding_window_func(
+                sliding_window_descriptor=window_descriptor,
+                agg_field_descriptors=agg_descriptors,
+                tz=tz,
+                parser=parser,
+                ast_evaluator=ast_evaluator,
+            )
+        )
+    else:
+        agg_df = df_copy.sort_values(by=[EVENT_TIME_ATTRIBUTE_NAME])
+        agg_df = _sliding_window_func(
+            df=agg_df,
+            sliding_window_descriptor=window_descriptor,
+            agg_field_descriptors=agg_descriptors,
+            tz=tz,
+            parser=parser,
+            ast_evaluator=ast_evaluator,
+        )
+
+    agg_df = agg_df.reset_index(drop=True)
+
+    # Compute the timestamp field with the given timestamp format from event
+    # time(window time).
+    if feature_view.timestamp_field is not None:
+        if feature_view.timestamp_format == "epoch":
+            agg_df[feature_view.timestamp_field] = agg_df[
+                EVENT_TIME_ATTRIBUTE_NAME
+            ].transform(lambda epoch_millis: int(epoch_millis / 1000))
+        elif feature_view.timestamp_format == "epoch_millis":
+            agg_df[feature_view.timestamp_field] = agg_df[EVENT_TIME_ATTRIBUTE_NAME]
+        else:
+            agg_df[feature_view.timestamp_field] = agg_df[
+                EVENT_TIME_ATTRIBUTE_NAME
+            ].transform(
+                lambda epoch_millis: datetime.fromtimestamp(
+                    epoch_millis / 1000
+                ).strftime(feature_view.timestamp_format)[:-3]
+            )
+
+    agg_df = agg_df.drop([EVENT_TIME_ATTRIBUTE_NAME], axis=1)
+    return agg_df
+
+
+def _get_sliding_window_func(
+    sliding_window_descriptor: SlidingWindowDescriptor,
+    agg_field_descriptors: Sequence[AggregationFieldDescriptor],
+    tz: tzinfo,
+    parser: ExprParser,
+    ast_evaluator: LocalAstEvaluator,
+) -> Callable[[pd.DataFrame], pd.DataFrame]:
+    def func(df: pd.DataFrame) -> pd.DataFrame:
+        return _sliding_window_func(
+            df,
+            sliding_window_descriptor,
+            agg_field_descriptors,
+            tz,
+            parser,
+            ast_evaluator,
+        )
+
+    return func
+
+
+def _sliding_window_func(
+    df: pd.DataFrame,
+    sliding_window_descriptor: SlidingWindowDescriptor,
+    agg_field_descriptors: Sequence[AggregationFieldDescriptor],
+    tz: tzinfo,
+    parser: ExprParser,
+    ast_evaluator: LocalAstEvaluator,
+) -> Optional[pd.DataFrame]:
+
+    df_copy = df.copy()
+    if df_copy.shape[0] <= 0:
+        return None
+
+    df_copy.sort_values(by=[EVENT_TIME_ATTRIBUTE_NAME], inplace=True)
+    df_copy[EVENT_TIME_ATTRIBUTE_NAME] = df_copy[EVENT_TIME_ATTRIBUTE_NAME].transform(
+        lambda t: _to_local_timestamp_mills(t, tz)
```

Comment:
Can you explain why we need this conversion?

Reply:
We need to convert the unix time to the local timestamp mills that describe the number of milliseconds since 1970-01-01 00:00:00 at the local timezone. We assign row based on the local timestamp millis instead of unix time so that the windows are aligned with 1970-01-01 00:00:00 at the current time zone.